botocore/client.py

```diff
@@ -170,33 +172,49 @@ def _load_service_model(self, service_name):
         return service_model
 
     def _get_client_args(self, service_model, region_name, is_secure,
-                         endpoint_url, verify, aws_access_key_id,
-                         aws_secret_access_key, aws_session_token):
+                         endpoint_url, verify, credentials,
+                         scoped_config):
         # A client needs:
         #
         # * serializer
         # * endpoint
         # * response parser
+        # * request signer
         protocol = service_model.metadata['protocol']
         serializer = botocore.serialize.create_serializer(
             protocol, include_validation=True)
-        creds = None
-        if aws_secret_access_key is not None:
-            creds = credentials.Credentials(
-                access_key=aws_access_key_id,
-                secret_key=aws_secret_access_key,
-                token=aws_session_token)
-        endpoint = self._endpoint_creator.create_endpoint(
+        event_emitter = copy.copy(self._event_emitter)
+        endpoint_creator = EndpointCreator(self._endpoint_resolver, region_name,
+                                           event_emitter, self._user_agent)
+        endpoint = endpoint_creator.create_endpoint(
             service_model, region_name, is_secure=is_secure,
             endpoint_url=endpoint_url, verify=verify,
-            credentials=creds,
             response_parser_factory=self._response_parser_factory)
         response_parser = botocore.parsers.create_parser(protocol)
+
+        # Get endpoint heuristic overrides before creating the
+        # request signer.
+        resolver = endpoint_creator.resolver
+        scheme = 'https' if is_secure else 'http'
+        endpoint_config = resolver.construct_endpoint(
+                service_model.endpoint_prefix,
+                region_name, scheme=scheme)
+        region_name = endpoint_config.get('properties', {}).get(
+            'credentialScope', {}).get('region', region_name)
+        signature_version = service_model.signature_version
+        if 'signatureVersion' in endpoint_config.get('properties', {}):
+            signature_version = endpoint_config['properties']\
+                                               ['signatureVersion']
+        signer = RequestSigner(service_model.service_name, region_name,
```

Comment:
I agree that it could be done here instead of as a customization, but I'm concerned that there might be other use cases for having access to the config in a customization. I'll think about it a bit, and if I don't come up with anything I'll move the logic here. We can always add the config to the events later if we need it, I suppose.